cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -0,0 +1,359 @@
+            let regL = ctx.input(iri, 0);
+            let regR = ctx.input(iri, 1);
+            let is64 = int_ty_to_is64(ty.unwrap());
+            let how = if op == Opcode::Iadd {
+                RMI_R_Op::Add
+            } else {
+                RMI_R_Op::Sub
+            };
+            ctx.emit(Inst::mov_r_r(true, regL, regD));
+            ctx.emit(Inst::alu_rmi_r(is64, how, RMI::reg(regR), regD));
+        }
+
+        Opcode::Ishl | Opcode::Ushr | Opcode::Sshr => {
+            // TODO: implement imm shift value into insn
+            let tySL = ctx.input_ty(iri, 0);
+            let tyD = ctx.output_ty(iri, 0); // should be the same as tySL
+            let regSL = ctx.input(iri, 0);
+            let regSR = ctx.input(iri, 1);
+            let regD = ctx.output(iri, 0);
+            if tyD == tySL && (tyD == types::I32 || tyD == types::I64) {
+                let how = match op {
+                    Opcode::Ishl => ShiftKind::Left,
+                    Opcode::Ushr => ShiftKind::RightZ,
+                    Opcode::Sshr => ShiftKind::RightS,
+                    _ => unreachable!(),
+                };
+                let is64 = tyD == types::I64;
+                let r_rcx = regs::rcx();
+                let w_rcx = Writable::<Reg>::from_reg(r_rcx);
+                ctx.emit(Inst::mov_r_r(true, regSL, regD));
+                ctx.emit(Inst::mov_r_r(true, regSR, w_rcx));
+                ctx.emit(Inst::shift_r(is64, how, None /*%cl*/, regD));
+            } else {
+                unimplemented!()
+            }
+        }
+
+        Opcode::Uextend | Opcode::Sextend => {
+            // TODO: this is all extremely lame, all because Mov{ZX,SX}_M_R
+            // don't accept a register source operand.  They should be changed
+            // so as to have _RM_R form.
+            // TODO2: if the source operand is a load, incorporate that.
+            let isZX = op == Opcode::Uextend;
+            let tyS = ctx.input_ty(iri, 0);
+            let tyD = ctx.output_ty(iri, 0);
+            let regS = ctx.input(iri, 0);
+            let regD = ctx.output(iri, 0);
+            ctx.emit(Inst::mov_r_r(true, regS, regD));
+            match (tyS, tyD, isZX) {
+                (types::I8, types::I64, false) => {
+                    ctx.emit(Inst::shift_r(true, ShiftKind::Left, Some(56), regD));
+                    ctx.emit(Inst::shift_r(true, ShiftKind::RightS, Some(56), regD));
+                }
+                _ => unimplemented!(),
+            }
+        }
+
+        Opcode::FallthroughReturn | Opcode::Return => {
+            for i in 0..ctx.num_inputs(iri) {
+                let src_reg = ctx.input(iri, i);
+                let retval_reg = ctx.retval(i);
+                ctx.emit(Inst::mov_r_r(true, src_reg, retval_reg));
+            }
+            // We don't generate the actual `ret` insn here (no way we could) since it first
+            // requires a prologue to be generated.  That's a job for the ABI machinery.
```

Comment:
I'm sure this is handled correctly, although I am not exactly sure how.  The new backend and register allocators are block-order invariant; they can be presented with the blocks in any order; they don't care.  After regalloc is done, a final block order is chosen, branches which would be fallthroughs are removed, empty blocks are removed, and the prologue/epilogue machinery identify the entry/exit blocks and add the relevant extra insns.

Reply:
I mean how does it know that this is an exit block if there is no `ret` emitted?

Reply:
You'd have to ask Chris or Benjamin for a definite answer, but .. I'm pretty sure it's because the higher level steering machinery in the new BE knows which CLIR-level blocks are exit blocks, so it knows which machine code blocks are exit blocks.

Reply:
Ha, interesting. Indeed it lacks the emission of `Ret`. This bug wouldn't be observable in spidermonkey since it uses the FallthroughRet, which means wasm returns in the middle of a function are translated into a jump to a single return block at the end of the function.

Reply:
Taking this back: the actual `Ret` opcode for this is generated in `machinst/lower.rs`, in `gen_retval_setup`.